lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -187,6 +187,117 @@ static SmallVector<int64_t> computeDimsOrderForMoveDim(int64_t srcDimInt,
+  Value middleDimProduct =
+      rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(1));
+  for (size_t i = 0; i < shape.size(); ++i) {
+    if (std::find(batchingDims.begin(), batchingDims.end(), i) ==
+            batchingDims.end() &&
+        std::find(contractingDims.begin(), contractingDims.end(), i) ==
+            contractingDims.end()) {
+      middleDimProduct =
+          rewriter.create<AtenMulIntOp>(loc, middleDimProduct, shape[i]);
+      otherDims.push_back(i);
+    }
+  }
+  int64_t otherDimsSize = otherDims.size();
+  if (!batchingDims.empty()) {
+    int64_t usedOtherDim = 0;
+    Value batchingDimProduct =
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(1));
+    int64_t batchingDimsRank = batchingDims.size();
+    for (int64_t i = 0; i < batchingDimsRank; ++i) {
+      batchingDimProduct =
+          rewriter.create<AtenMulIntOp>(loc, batchingDimProduct,
+                                        shape[batchingDims[i]]);
+      if (batchingDims[i] != i) {
+        Value batchingDim =
+            rewriter.create<ConstantIntOp>(loc,
+                                           rewriter.getI64IntegerAttr(
+                                               batchingDims[i]));
+        Value indexDim = rewriter.create<ConstantIntOp>(
+            loc, rewriter.getI64IntegerAttr(otherDims[usedOtherDim]));
+        inputTensor = rewriter.create<AtenTransposeIntOp>(
+            loc, op->getResultTypes(), inputTensor, batchingDim, indexDim);
+        usedOtherDim += 1;
+      }
+    }
+    finalShape.push_back(batchingDimProduct);
+  }
+  if (!otherDims.empty())
+    finalShape.push_back(middleDimProduct);
+  if (!contractingDims.empty()) {
+    int64_t usedOtherDim = 1;
+    int64_t rank = tokens.size();
+    Value contractingDimProduct =
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(1));
+    int64_t contractingDimsRank = contractingDims.size();
+    for (int64_t i = contractingDimsRank - 1; i > -1; --i) {
+      contractingDimProduct =
+          rewriter.create<AtenMulIntOp>(loc, contractingDimProduct,
+                                        shape[contractingDims[i]]);
+      if (contractingDims[i] != rank - contractingDimsRank + i) {
+        Value contractingDim =
+            rewriter.create<ConstantIntOp>(loc,
+                                           rewriter.getI64IntegerAttr(
+                                               contractingDims[i]));
+        Value indexDim = rewriter.create<ConstantIntOp>(
+            loc, rewriter.getI64IntegerAttr(
+                     otherDims[otherDimsSize - usedOtherDim]));
+        inputTensor = rewriter.create<AtenTransposeIntOp>(
+            loc, op->getResultTypes(), inputTensor, contractingDim, indexDim);
+        usedOtherDim += 1;
+      }
+    }
+    finalShape.push_back(contractingDimProduct);
+  }
+  return inputTensor;
+}
+
+static Value createReshapedTensor(PatternRewriter &rewriter, Location loc,
```

Comment:
nit: I don't think this is really needed, since it does not save that many lines of code